c10/util/complex.h

```diff
@@ -252,8 +252,10 @@ struct alignas(sizeof(T) * 2) complex {
     T ai = imag_;
     U br = rhs.real();
     U bi = rhs.imag();
-    auto abs_br = std::abs(br);
-    auto abs_bi = std::abs(bi);
+    // not using std::abs because it is not a constexpr and
+    // can cause an error in windows
+    auto abs_br = br < 0 ? -br : br;
+    auto abs_bi = bi < 0 ? -bi : bi;
```

Comment:
Can you check in godbolt whether this generates the same assembly? Otherwise, we'll need to choose between the two implementations via a compiler flag

Reply:
Not sure how to check the assembly on godbolt. I tried the following:
```
#include <cmath>
float abs1(float f1) {return std::abs(f1);}
float abs2(float f2) {return f2 < 0 ? -f2 : f2;}
```
and it just gives me this:
```
# Compilation provided by Compiler Explorer at https://godbolt.org/
abs1(float):
        push    rbp
        mov     rbp, rsp
        sub     rsp, 16
        movss   DWORD PTR [rbp-4], xmm0
        mov     eax, DWORD PTR [rbp-4]
        movd    xmm0, eax
        call    std::abs(float)
        leave
        ret
abs2(float):
        push    rbp
        mov     rbp, rsp
        movss   DWORD PTR [rbp-4], xmm0
        pxor    xmm0, xmm0
        comiss  xmm0, DWORD PTR [rbp-4]
        jbe     .L11
        movss   xmm0, DWORD PTR [rbp-4]
        movss   xmm1, DWORD PTR .LC2[rip]
        xorps   xmm0, xmm1
        jmp     .L9
.L11:
        movss   xmm0, DWORD PTR [rbp-4]
.L9:
        pop     rbp
        ret
.LC0:
        .long   2147483647
        .long   0
        .long   0
        .long   0
.LC2:
        .long   -2147483648
        .long   0
        .long   0
        .long   0
```

Reply:
I checked the source code for `std::abs` [here](https://github.com/gcc-mirror/gcc/blob/47465fff9738b08796f4b52570fec49bbdbf3e57/libstdc%2B%2B-v3/include/bits/std_abs.h#L83-L98) and it's basically `__x >= 0 ? __x : -__x;`

Reply:
You need to enable the optimisations:
https://godbolt.org/z/1njcGsMPE
Try implementing it as GCC does then. Probably your current implementation has issues with `-0.0` which is why it can't optimised into a call to `andps`.

Reply:
Nah, it doesn't cut it. See that the implementation that you found is for int types. For float types, it dispatches to `__builtin_fabsf`, which then dispatches to the faster instruction.

You can do two things:
- The easy way out is to simply add a flag and based on whether we are on MSVC, dispatch to `at::abs()` or `x >= 0 ? x : -x`
- The tryharder way, see how MSVC implements the fast path and try to use that implementation, which should generate the same assembly.

Reply:
Thanks, @lezcano. I'll just take the easy way.
BTW, if compiled with `-Ofast`, it reduces to the same assembly (https://godbolt.org/z/a7j5MMnMG). But it only works if pytorch compiles with `-Ofast` flag.

Reply:
Yeah, `Ofast` cuts too many corners, you cannot use it in a numeric library. Julia had some real issues with that.
What that means in practice is that `std::abs` is not equivalent to `x >= 0 ? x : -x` in some corner case ([`-0.0`, I'm looking at you](https://godbolt.org/z/rofvf317j)). Now, that's little detail is alright in this case really.